docs/configuration.md

```diff
@@ -326,4 +326,209 @@ deployment_config = DeploymentConfig(
+New API is optional and there is no need to migrate existing code from
+`biglfow.Configuration`. However it betters plays with typehinting, enables
+autocompletion in IDEs and gives more flexibility.
+
+Each configuration is declared as a subclass of `bigflow.konfig.Konfig`.
+Properties are declared as attributes, optionally with type hints.
+Classes may inherit from each other and share common properties.  Standard
+python inheritance rules are used to resolve overriden properties.
+Custom methods may be added to classes, however it is not recommended.
+
+```python
+import bigflow.konfig
+
+class CommonConf(bigflow.konfig.Konfig):
+    project: str                    # type hint
+    region: str = 'europe-west1'    # type hint + default value
+    machine_type = 'n2-standard-4'  # only value (type is `typing.Any`)
+    num_workers = 4                 # not only strings are allowed
+
+class DevConf(CommonConf):
+    project = "dev-project"
+
+class ProdConf(CommonConf):
+    project = "prod-project"
+    num_workers = 20
+```
+
+Config which corresponds to the current environment may be created by function `resolve_konfig`.
+It takes dictionary in form of {"environment" -> "config class"} and returns instance of the config.
+
+```python
+config = bigflow.konfig.resolve_konfig(
+    {  # mapping of all configs
+        'dev': DevConf,
+        'prod': ProdConf,
+    },
+    default="dev",  # when no environment is porvided via CLI
+)
+print(config.project)      # => "dev-project"
+print(config.num_workers)  # => 4
+```
+
+Config instance may be used as a dict:
+
+```python
+print(config['project'])
+print(config.keys())
+print(dict(config))
+```
+
+Also config may be created manullay in tests or interactive shell:
+
+```python
+dev_config = DevConf()
+print(dev_config.num_workers)
+```
+
+Attributes may be overwritten by passing parameters to the class:
+
+```python
+patched_dev_config = DevConf(num_instances=10)
+```
+
+### String interpolation
+
+By default string properties are *not* interpolated (it differs from `bigflow.Configuration`).
+You need to be explicitly wrap interpolation pattern with function `bigflow.konfig.expand()`:
+
+```python
+from bigflow.konfig import expand
+
+class MyConfig(bigflow.konfig.Konfig):
+
+    docker_repository_project = "my-shared-docker-project-id"
+
+    # no 'expand()' function - string is *not* inerpolcatied
+    docker_repository_raw = "eu.gcr.io/{docker_repository_project}/my-analytics"
+
+    # interpolation is enabled
+    docker_repository = expand("eu.gcr.io/{docker_repository_project}/my-analytics")
+
+
+config = MyConfig()
+print(config.docker_repository_raw)  # => eu.gcr.io/{docker_repository_project}/my-analytics
+print(config.docker_repository)      # => eu.gcr.io/my-shared-docker-project-id/my-analytics
+```
+
+### Reading from environment variables
+
+Class-based configs system does not automatically read properties from OS environment.
```

Comment:
Why it does not automatically read them?

Reply:
Mostly because
> Explicit is better than implicit.

It might be hard to code reader to recognise that some property comes from environment.

Reply:
I opt for explicit solution